### DISCUSSION TITLE
Anything about architecture and code quality

### DISCUSSION THREAD

**Discussion by joooye34** [toeverything/AFFiNE#4534] [General]
The code of AFFiNE need a new refactor to have a reasonable architecture. Anything you want to disclose, or any shit code you want to talk, pls write down here. Let us make it great.

**Comment by pengx17** [comment]
The env parsing/configuration of stable/beta/canary is scattered in different parts of the project. It's best to organize it and achieve a single source of truth.
The meaning of "stable/beta/canary/internal" behind "BUILD_TYPE" itself refers to the versions of the client, but now it is also used for configuration in the cloud environment, which feels a bit messy. We also have "BUILD_TYPE_OVERRIDE", which makes it even more strange.

**Reply by JimmFly** [reply]
Can you tell me how to get the current BUILD_TYPE in the component?

**Reply by pengx17** [reply]
> Can you tell me how to get the current BUILD_TYPE in the component?

I think it is not exposed into the app yet but you can add one to environment.

**Comment by pengx17** [comment]
IMO affine should not have different workspace types. Right now workspace is categorized into cloud/local/public etc, but they are not either one or another relationship among them. E.g, a cloud workspace is also a local workspace. What matters most is the provider configs for syncing data between different data sources.

**Comment by joooye34** [comment]
The directory structure is confused, product code and development tool code are mixed together. It is hard to quickly determine where the code should be written.

**Comment by pengx17** [comment]
Provider performance issue. Especially on Edgeless that writes in every frame to different datasources.
We have some basic solution on merging updates, but it seems that it is not sufficient.
